discovery-provider/src/tasks/index_user_bank.py

```diff
@@ -203,10 +226,255 @@ def process_create_userbank_instruction(
+        signature=tx_sig,
+        transaction_type=USDCTransactionType.purchase_content,
+        method=USDCTransactionMethod.receive,
+        transaction_created_at=timestamp,
+        change=Decimal(balance_changes[receiver_account]["change"]),
+        balance=Decimal(balance_changes[receiver_account]["post_balance"]),
+    )
+    session.add(usdc_tx_received)
+    logger.debug(
+        f"index_user_bank.py | Creating usdc_tx_history received tx for purchase {usdc_tx_received}"
+    )
+
+
+def validate_and_index_purchase(
+    session: Session,
+    receiver_user_id: Union[int, None],
+    receiver_account: str,
+    sender_user_id: int,
+    sender_account: str,
+    balance_changes: dict[str, BalanceChange],
+    purchase_metadata: Union[PurchaseMetadataDict, None],
+    slot: int,
+    timestamp: datetime.datetime,
+    tx_sig: str,
+):
+    """Checks if the transaction is a valid purchase and if so creates the purchase record. Otherwise, indexes a transfer."""
+    if purchase_metadata is not None and validate_purchase(
+        purchase_metadata=purchase_metadata,
+        balance_changes=balance_changes,
+        sender_account=sender_account,
+        receiver_account=receiver_account,
+    ):
+        index_purchase(
+            session=session,
+            receiver_user_id=receiver_user_id,
+            receiver_account=receiver_account,
+            sender_user_id=sender_user_id,
+            sender_account=sender_account,
+            balance_changes=balance_changes,
+            purchase_metadata=purchase_metadata,
+            slot=slot,
+            timestamp=timestamp,
+            tx_sig=tx_sig,
+        )
+    else:
+        # Both non-purchases and invalid purchases will be treated as transfers
+        usdc_tx_sent = USDCTransactionsHistory(
+            user_bank=sender_account,
+            slot=slot,
+            signature=tx_sig,
+            transaction_type=USDCTransactionType.transfer,
+            method=USDCTransactionMethod.send,
+            transaction_created_at=timestamp,
+            change=Decimal(balance_changes[sender_account]["change"]),
+            balance=Decimal(balance_changes[sender_account]["post_balance"]),
+        )
+        logger.debug(f"index_user_bank.py | Creating transfer sent tx {usdc_tx_sent}")
+        session.add(usdc_tx_sent)
+        usdc_tx_received = USDCTransactionsHistory(
+            user_bank=receiver_account,
+            slot=slot,
+            signature=tx_sig,
+            transaction_type=USDCTransactionType.transfer,
+            method=USDCTransactionMethod.receive,
+            transaction_created_at=timestamp,
+            change=Decimal(balance_changes[receiver_account]["change"]),
+            balance=Decimal(balance_changes[receiver_account]["post_balance"]),
+        )
+        session.add(usdc_tx_received)
+        logger.debug(
+            f"index_user_bank.py | Creating transfer received tx {usdc_tx_received}"
+        )
+
+
+def index_user_tip(
```

Comment:
nit: `index_audio_user_tip`? Feels like it would be nice to include USDC/AUDIO in the fn names since this file has fns that operate on both types

Reply:
I like to think of the token mint as an impl detail - audio tips are just tips, usdc purchases are just purchases? can always rename these when that's not true